# Guided Capstone Step 4. Pre-Processing and Training Data Development

**The Data Science Method**  


1.   Problem Identification 


2.   Data Wrangling 
  
 
3.   Exploratory Data Analysis   

4.   **Pre-processing and Training Data Development**  
 * Create dummy or indicator features for categorical variables
  * Standardize the magnitude of numeric features
  * Split into testing and training datasets
  * Apply scaler to the testing set
5.   Modeling 
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

**<font color='teal'> Start by loading the necessary packages as we did in step 3 and printing out our current working directory just to confirm we are in the correct project directory. </font>**

In [1]:
#load python packages
import os
import pandas as pd
import datetime
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np
%matplotlib inline


In [2]:
pwd

'C:\\Users\\ffgfggg\\Downloads\\Boot Camps\\springboard\\1582673773_GuidedCapstone_Step2.ipynb'

**<font color='teal'>  Load the csv file you created in step 3, remember it should be saved inside your data subfolder and print the first five rows.</font>**

In [3]:
df=pd.read_csv('step3_output.csv')

## Create dummy features for categorical variables

**<font color='teal'> Create dummy variables for `state`. Add the dummies back to the dataframe and remove the original column for `state`. </font>**

Hint: you can see an example of how to execute this in Aiden's article on preprocessing [here](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-pre-processing-and-training-data-development-fd2d75182967). 

In [4]:
print(df.columns)
print(df.shape)
print(df.head())
df.drop('Unnamed: 0',axis=1,inplace=True)


Index(['Unnamed: 0', 'Name', 'state', 'summit_elev', 'vertical_drop', 'trams',
       'fastEight', 'fastSixes', 'fastQuads', 'quad', 'triple', 'double',
       'surface', 'total_chairs', 'Runs', 'TerrainParks', 'LongestRun_mi',
       'SkiableTerrain_ac', 'Snow Making_ac', 'daysOpenLastYear', 'yearsOpen',
       'averageSnowfall', 'AdultWeekday', 'AdultWeekend', 'projectedDaysOpen',
       'NightSkiing_ac', 'clusters'],
      dtype='object')
(172, 27)
   Unnamed: 0                           Name       state  summit_elev  \
0           2               Hilltop Ski Area      Alaska         2090   
1           4            Sunrise Park Resort     Arizona        11100   
2           5  Yosemite Ski & Snowboard Area  California         7800   
3           8         Boreal Mountain Resort  California         7700   
4           9                    Dodge Ridge  California         8200   

   vertical_drop  trams  fastEight  fastSixes  fastQuads  quad  ...  \
0            294      0        0.0

In [5]:
df = pd.concat([df, pd.get_dummies(df.state)], axis=1)
print(df.shape)
print(df.head())

(172, 61)
                            Name       state  summit_elev  vertical_drop  \
0               Hilltop Ski Area      Alaska         2090            294   
1            Sunrise Park Resort     Arizona        11100           1800   
2  Yosemite Ski & Snowboard Area  California         7800            600   
3         Boreal Mountain Resort  California         7700            500   
4                    Dodge Ridge  California         8200           1600   

   trams  fastEight  fastSixes  fastQuads  quad  triple  ...  Rhode Island  \
0      0        0.0          0          0     0       1  ...             0   
1      0        0.0          0          1     2       3  ...             0   
2      0        0.0          0          0     0       1  ...             0   
3      0        0.0          0          1     1       3  ...             0   
4      0        0.0          0          0     1       2  ...             0   

   South Dakota  Tennessee  Utah  Vermont  Virginia  Washington 

In [6]:
df.drop('state',axis=1,inplace=True)
print(df.shape)
print(df.head())

(172, 60)
                            Name  summit_elev  vertical_drop  trams  \
0               Hilltop Ski Area         2090            294      0   
1            Sunrise Park Resort        11100           1800      0   
2  Yosemite Ski & Snowboard Area         7800            600      0   
3         Boreal Mountain Resort         7700            500      0   
4                    Dodge Ridge         8200           1600      0   

   fastEight  fastSixes  fastQuads  quad  triple  double  ...  Rhode Island  \
0        0.0          0          0     0       1       0  ...             0   
1        0.0          0          1     2       3       1  ...             0   
2        0.0          0          0     0       1       3  ...             0   
3        0.0          0          1     1       3       1  ...             0   
4        0.0          0          0     1       2       5  ...             0   

   South Dakota  Tennessee  Utah  Vermont  Virginia  Washington  \
0             0      

In [7]:
print(df.describe().T)

                   count         mean          std    min      25%     50%  \
summit_elev        172.0  3637.656977  3223.022158  315.0  1245.00  2000.0   
vertical_drop      172.0   841.447674   584.594889  175.0   391.25   650.0   
trams              172.0     0.000000     0.000000    0.0     0.00     0.0   
fastEight          172.0     0.000000     0.000000    0.0     0.00     0.0   
fastSixes          172.0     0.000000     0.000000    0.0     0.00     0.0   
fastQuads          172.0     0.238372     0.588520    0.0     0.00     0.0   
quad               172.0     0.540698     0.703854    0.0     0.00     0.0   
triple             172.0     1.209302     1.234175    0.0     0.00     1.0   
double             172.0     1.720930     1.472168    0.0     1.00     1.0   
surface            172.0     2.261628     1.445353    0.0     1.00     2.0   
total_chairs       172.0     5.970930     2.768721    0.0     4.00     5.0   
Runs               172.0    29.715116    17.443148    7.0    16.

In [8]:
print(df.sum())


Name                 Hilltop Ski AreaSunrise Park ResortYosemite Sk...
summit_elev                                                     625677
vertical_drop                                                   144729
trams                                                                0
fastEight                                                            0
fastSixes                                                            0
fastQuads                                                           41
quad                                                                93
triple                                                             208
double                                                             296
surface                                                            389
total_chairs                                                      1027
Runs                                                              5111
TerrainParks                                                       303
Longes

## Standardize the magnitude of numeric features

**<font color='teal'> Using sklearn preprocessing standardize the scale of the features of the dataframe except the name of the resort which we done't need in the dataframe for modeling, so it can be droppped here as well. Also, we want to hold out our response variable(s) so we can have their true values available for model performance review. Let's set `AdultWeekend` to the y variable as our response for scaling and modeling. Later we will go back and consider the `AdultWeekday`, `dayOpenLastYear`, and `projectedDaysOpen`. For now leave them in the development dataframe. </font>**

In [9]:
# first we import the preprocessing package from the sklearn library
from sklearn import preprocessing

# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X= df.drop(['Name','AdultWeekend'], axis=1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y= df.AdultWeekend 

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler().fit(X)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled=scaler.transform(X) 

## Split into training and testing datasets

**<font color='teal'> Using sklearn model selection import train_test_split, and create a 75/25 split with the y = `AdultWeekend`. We will start by using the adult weekend ticket price as our response variable for modeling.</font>**

In [10]:
# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)

Here we start the actual modeling work. First let's fit a multiple linear regression model to predict the `AdultWeekend` price.

# Guided Capstone Step 5. Modeling


This is the fifth step in the Data Science Method. In the previous steps you cleaned and prepared the datasets. Now it's time to get into the most exciting part: modeling! In this exercise, you'll build three different models and compare each model's performance. In the end, you'll choose the best model for demonstrating insights to Big Mountain management.



### **The Data Science Method**  


1.   Problem Identification 

2.   Data Wrangling 
  
3.   Exploratory Data Analysis 
 
4.   Pre-processing and Training Data Development

5.   **Modeling**
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

## Fit Models with a Training Dataset

**<font color='teal'> Using sklearn, fit the model on your training dataset.</font>**

#### Model 1

In [11]:
#all first model set
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)

**<font color='teal'> Predict on the testing dataset and score the model performance with the y_test set and the y-pred values. The explained variance is a measure of the variation explained by the model. This is also known as the R-squared value. </font>**

Hint: you will have to use the `predict()` method here as it's used in this [DSM article](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-modeling-56b4233cad1b) about modeling. 

In [12]:
# Make a variable called y_pred and assign it the result of calling predict() on our model variable with parameter X_test
y_pred = model.predict(X_test)

## Review Model Outcomes — Iterate over additional models as needed

In [13]:
# You might want to use the explained_variance_score() and mean_absolute_error() metrics.
# To do so, you will need to import them from sklearn.metrics. 
# You can plug y_test and y_pred into the functions to evaluate the model
from sklearn.metrics import explained_variance_score, mean_absolute_error
print(explained_variance_score(y_test, y_pred))

-4.06138750714297e+21


In [14]:
print(mean_absolute_error(y_test, y_pred))

242061106889.92587


**<font color='teal'> Print the intercept value from the linear model. </font>**

Hint: our linear regression model `lm` has an attribute `intercept_` for the intercept

In [15]:
print(lm.intercept_)

-60515276664.12326


**<font color='teal'> The intercept is the mean `AdultWeekend` price for all the resorts given the other characteristics. The addition or subtraction of each of the coefficient values in the regression are numeric adjustments applied to the intercept to provide a particular observation's value for the resulting `AdultWeekend` value. Also, because we took the time to scale our x values in the training data, we can compare each of the coeeficients for the features to determine the feature importances. Print the coefficient values from the linear model and sort in descending order to identify the top ten most important features.</font>** 


Hint: make sure to review the absolute value of the coefficients, because the adjustment may be positive or negative, but what we are looking for is the magnitude of impact on our response variable.

In [16]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
coefficients=pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient'])
print(coefficients.sort_values('Coefficient',ascending=False))


                    Coefficient
fastEight          4.338382e+13
trams              1.106195e+13
total_chairs       5.915127e+12
fastSixes          3.419268e+12
double             3.145156e+12
surface            3.087868e+12
triple             2.636705e+12
New York           1.668347e+12
quad               1.503721e+12
Michigan           1.468402e+12
California         1.325868e+12
New Hampshire      1.273383e+12
Pennsylvania       1.273383e+12
fastQuads          1.257320e+12
Wisconsin          1.217887e+12
Vermont            1.096015e+12
Minnesota          1.096015e+12
New Mexico         9.549470e+11
Idaho              9.549470e+11
Maine              8.743651e+11
Connecticut        8.743651e+11
Wyoming            8.743651e+11
Massachusetts      7.843939e+11
Colorado           7.843939e+11
Ohio               7.843939e+11
North Carolina     6.813238e+11
West Virginia      6.813238e+11
Washington         6.813238e+11
Utah               6.813238e+11
Oregon             6.813238e+11
Iowa    

**<font color='teal'>You should see that the top ten important features are different states. However, the state is not something the managers at the Big Mountain Resort can do anything about. Given that we care more about actionable traits associated with ticket pricing, rebuild the model without the state features and compare the results. </font>**

Hint: Try to construct another model using exactly the steps we followed above. 

#### Model 2

In [17]:
df=pd.read_csv('step3_output.csv')

# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X= df.drop(['Name','AdultWeekend','state'], axis=1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y= df.AdultWeekend 

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler().fit(X)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled=scaler.transform(X) 

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)

lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)

# Make a variable called y_pred and assign it the result of calling predict() on our model variable with parameter X_test
y_pred = model.predict(X_test)

print(explained_variance_score(y_test, y_pred))
print(mean_absolute_error(y_test, y_pred))
print(lm.intercept_)

# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
coefficients=pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient'])
print(coefficients.sort_values('Coefficient',ascending=False).head(10))

0.8354076184707633
5.90295085867451
57.02736106593792
                   Coefficient
AdultWeekday          9.490599
summit_elev           3.452938
clusters              2.800400
daysOpenLastYear      2.258535
vertical_drop         1.922233
averageSnowfall       1.495699
TerrainParks          1.443602
projectedDaysOpen     1.381768
LongestRun_mi         1.268194
Snow Making_ac        0.992422


**<font color='teal'> When reviewing our new model coefficients, we see `summit_elev` is now in the number two spot. This is also difficult to change from a management prespective and highly correlated with `base_elev` and `vertical_drop`.  This time, rebuild the model without the state features and without the `summit_elev` and without `base_elev`and compare the results. </font>**

#### Model 3

In [18]:
df=pd.read_csv('step3_output.csv')

# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X= df.drop(['Name','AdultWeekend','state', 'summit_elev'], axis=1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y= df.AdultWeekend 

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler().fit(X)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled=scaler.transform(X) 

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)

lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)

# Make a variable called y_pred and assign it the result of calling predict() on our model variable with parameter X_test
y_pred = model.predict(X_test)

print(explained_variance_score(y_test, y_pred))
print(mean_absolute_error(y_test, y_pred))
print(lm.intercept_)

# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
coefficients=pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient'])
print(coefficients.sort_values('Coefficient',ascending=False).head(10))

0.8468741268430149
5.779559773949307
56.99846326944172
                   Coefficient
AdultWeekday          9.274632
averageSnowfall       2.494497
clusters              2.007195
daysOpenLastYear      1.817794
vertical_drop         1.548945
LongestRun_mi         1.475045
Snow Making_ac        1.428658
SkiableTerrain_ac     1.385521
TerrainParks          1.374885
projectedDaysOpen     0.836120


## Identify the Final Model

**<font color='teal'> Review the model performances in the table below and choose the best model for proving insights to Big Mountain management about what features are driving ski resort lift ticket prices. Type your choice in the final markdown cell — you will discuss this selection more in the next step of the guided casptone. </font>**

| Model | Explained Variance| Mean Absolute Error|Features Dropped|
| --- | --- | --- | --- |
| Model 1. | 0.745? | 6.44? |-|
| Model 2. | 0.835 | 5.90 |'state'|
| Model 3. | 0.847 | 5.78 |'state','summit_elev','base_elev'|

Model Selection:

In [19]:
#Model 3 looks the best, right?